json-transport/src/lib.rs

```diff
@@ -20,37 +20,27 @@ use std::{
     task::{Context, Poll},
 };
 use tokio::codec::{length_delimited::LengthDelimitedCodec, Framed};
-use tokio_io::{AsyncRead, AsyncWrite};
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio::net::{TcpListener, TcpStream};
+use tokio_net::ToSocketAddrs;
 use tokio_serde_json::*;
-use tokio_tcp::{TcpListener, TcpStream};
 
 /// A transport that serializes to, and deserializes from, a [`TcpStream`].
-pub struct Transport<S: AsyncWrite, Item, SinkItem> {
-    inner: Compat01As03Sink<
-        ReadJson<WriteJson<Framed<S, LengthDelimitedCodec>, SinkItem>, Item>,
-        SinkItem,
-    >,
-}
-
-impl<S: AsyncWrite, Item, SinkItem> Transport<S, Item, SinkItem> {
-    unsafe_pinned!(
-        inner:
-            Compat01As03Sink<
-                ReadJson<WriteJson<Framed<S, LengthDelimitedCodec>, SinkItem>, Item>,
-                SinkItem,
-            >
-    );
+#[pin_project]
+pub struct Transport<S, Item, SinkItem> {
+    #[pin]
+    inner: ReadJson<WriteJson<Framed<S, LengthDelimitedCodec>, SinkItem>, Item>,
 }
 
 impl<S, Item, SinkItem> Stream for Transport<S, Item, SinkItem>
 where
-    S: AsyncWrite + AsyncRead,
+    S: AsyncWrite + AsyncRead + Unpin,
```

Comment:
It is.

Reply:
Can you add a comment above this code that explains why Unpin is needed?

Reply:
`S` is not `Stream` without `Unpin` bound. Do you want that note in the code?

Reply:
Do you mean `ReadJson<WriteJson<Framed<S, LengthDelimitedCodec>, SinkItem>, Item>` is not `Stream` without `S: Unpin`? What I'm looking for is an understanding of where this limitation comes from, so that if in the future I or another motivated individual could explore relaxing this restriction.

I found the source ([link](https://docs.rs/tokio/0.2.0-alpha.6/tokio/codec/struct.Framed.html#impl-Stream)):

```rust
impl<T, U> Stream for Framed<T, U>
where
    T: AsyncRead + Unpin,
    U: Decoder + Unpin, 
```

Ideally, there'd be an issue to track on the tokio github repo. Maybe [this one](https://github.com/tokio-rs/tokio/issues/1272)?

Reply:
I think so. I never looked closely though, just made sure that it compiled. I can add a TODO note with a link to the issue if you think it is necessary.

Reply:
Yeah, let's do that. Thanks!